### DISCUSSION TITLE
Board randomly stuck during game playback

### DISCUSSION THREAD

**Discussion by und3f** [BattlesnakeOfficial/feedback#251] [Bug Reports]
Sometimes board randomly stuck during game playback for the game that ended before playback.

Attached recording, you may see board stuck on second run right on the last move, and the board for some reason is showing the preloader (doesn't doesn't change over time).

![battlesnake-stuck](https://github.com/BattlesnakeOfficial/feedback/assets/168848/7f332b16-2f3a-4882-8581-c011a9544eab)

**Comment by und3f** [comment]
During investigation of the bug root cause found that board reconnects to the engine multiple times as it did not receive all frames (Img.1, Img.2).

After discussion with @bvanvugt tried connections to events websocket from different places. 
Connected tested using `websockat` with `jq` for better output.
```bash
./websocat -E wss://engine.battlesnake.com/games/18059b87-8013-4003-b0bd-55f8a06d84ba/events | jq '.Data.Status, .Data.Turn'
```

- Home place mostly fail to receive all frames.
  -  Stops during 332-336 frames. Rarely on 337 frame
- OCI instance fail to receive all frames.
  -  Similar to Home, but it receives all frames often.
- Google cloud instance always receive all frames on each call.
  -  It always receive all frames.

Img. 1. First events websocket connection:
![image](https://github.com/BattlesnakeOfficial/feedback/assets/168848/7b92ff1b-5faa-47fe-8658-4a3749f80c80)

Img. 2. Fourth events websocket connection that failed to obtain all frames:
![image](https://github.com/BattlesnakeOfficial/feedback/assets/168848/4bfbd9a7-c30d-4dee-ae39-a13c072f80b4)

**Comment by bvanvugt** [comment]
Okay! I got a chance to do some digging into this, and found a couple engine tweaks that seem to have made a large improvement. Here are the details...

* The game engine uses [Gorilla Websockets](https://github.com/gorilla/websocket) to power websocket connections with the game board
* I was able to confirm that engine is in fact **writing all frames** to the socket, before actively closing the connection. This typically took less than 1,000ms, even for very long games.
* In certain situations, the board client was **not receiving all frames** before the connection was closed. The chances of this occurring seem to be related to physical distance between the board client and the engine servers, which are hosted in US-WEST.
* This lead me to believe the core issue was related to the write buffers not flushing completely before the connection was being closed.
* I made two tweaks to the way these connections are handled:
  * The write buffer on these connections was very large by default, so I reduced it to 1KB
  * I added a flush grace period before closing the connection (this is recommended as a best practice by the package)

These two changes combined seem to have resolved the issue, with @und3f reporting 100% frames received on all tests afterwards 👍 

It remains a goal of mine to open source these parts of the game engine to allow for stronger collaboration -- I hate that some of these issues can only be worked on by select few individuals with restricted free time :)

**Reply by und3f** [reply]
Thanks for digging it! The game board works smoothly now and playing games is pleasant! I can confirm that now I have 100% success rate on obtaining all frames during first WebSocket connection.
